plugins/ccgx/fu-ccgx-pure-hid-device.c

```diff
@@ -0,0 +1,559 @@
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(device),
+				      buf[0],
+				      buf,
+				      sizeof(buf),
+				      FU_CCGX_PURE_HID_DEVICE_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_NONE,
+				      error)) {
+		g_prefix_error(error, "write row command error: ");
+		return FALSE;
+	}
+	g_debug("After write row");
+
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_pure_hid_flash_firmware_image(FuDevice *device,
+				      const guint8 *fw_bytes,
+				      gsize fw_len,
+				      guint16 start_row,
+				      guint16 metadata_row,
+				      guint8 fw_img_no,
+				      GError **error)
+{
+	gsize rows;
+
+	// TODO: Probably not required
+	if (!fu_ccgx_pure_hid_get_fw_info_wrapper(device, NULL, error)) {
+		g_prefix_error(error, "get fw info error: ");
+		return FALSE;
+	}
+
+	g_assert(fw_len % ROW_SIZE == 0);
+	rows = fw_len / ROW_SIZE;
+
+	for (gsize row_no = 0; row_no < rows; row_no++) {
+		const guint8 *row = fw_bytes + row_no * ROW_SIZE;
+		if (row_no == 0) {
+			g_debug("Writing first row");
+		}
+		if (row_no == rows - 1) {
+			g_debug("Writing metadata row");
+			// Last row is the metadata row. It's at a different offset
+			if (!fu_ccgx_pure_hid_write_row(device, metadata_row, row, error))
+				return FALSE;
+		} else {
+			g_debug("Writing row #%lu 0x%04lx", row_no, start_row + row_no);
+			if (!fu_ccgx_pure_hid_write_row(device, start_row + row_no, row, error))
+				return FALSE;
+		}
+	}
+
+	// TODO: Doing this and reset by themselves (with magic unlock) doesn't
+	// switch to the alternative image. Seems we always need to flash in
+	// order to switch.
+	g_debug("before bootswitch");
+	if (!fu_ccgx_pure_hid_command(device, 0x04, fw_img_no, error)) {
+		g_prefix_error(error, "bootswitch command error: ");
+		return FALSE;
+	}
+	g_debug("After bootswitch");
+
+	g_debug("before reset");
+	if (!fu_ccgx_pure_hid_command(device,
+				      CCGX_HID_CMD_JUMP,
+				      CY_PD_DEVICE_RESET_CMD_SIG,
+				      error)) {
+		g_prefix_error(error, "reset command error: ");
+		return FALSE;
+	}
+	g_debug("After reset");
+
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_pure_hid_device_write_firmware(FuDevice *device,
+				       FuFirmware *firmware,
+				       FuProgress *progress,
+				       FwupdInstallFlags flags,
+				       GError **error)
+{
+	FuCcgxPureHidDevice *self = FU_CCGX_PURE_HID_DEVICE(device);
+	g_autoptr(GBytes) fw = NULL;
+	gsize fw_total_size;
+	gsize fw_size;
+	const guint8 *fw_bytes;
+	const guint8 *fw1_binary;
+	const guint8 *fw2_binary;
+	g_debug("write firmware");
+
+	/* get blob and read */
+	fw = fu_firmware_get_bytes(FU_FIRMWARE(firmware), error);
+	if (fw == NULL)
+		return FALSE;
+
+	fw_bytes = g_bytes_get_data(fw, NULL);
+	fw_total_size = g_bytes_get_size(fw);
+	fw_size = fw_total_size / 2;
+
+	// TODO: Do this in prepare_firmware
+	// Make sure there are two firmware images composed of full rows
+	g_assert(fw_total_size % 2 * ROW_SIZE == 0);
+	fw1_binary = fw_bytes;
+	fw2_binary = &fw_bytes[fw_size];
```

Comment:
is this how you got the firmware from the ODM, or did you manually "stick together" two files into one binary payload?

Reply:
I manually concatenated them. Will switch to CYACD format.

Reply:
Cool. This is why each bank has it's own guid btw; so you can match the correct firmware for each bank in the metainfo file.